sound/soc/sof/intel/hda-stream.c

```diff
@@ -119,13 +119,33 @@ int hda_dsp_stream_setup_bdl(struct snd_sof_dev *sdev,
 	int remain, ioc;
 
 	period_bytes = hstream->period_bytes;
-	dev_dbg(sdev->dev, "period_bytes:0x%x\n", period_bytes);
-	if (!period_bytes)
+	dev_dbg(sdev->dev, "period_bytes: %#x, bufsize: %#x\n", period_bytes,
+		hstream->bufsize);
+
+	/*
+	 * HDA spec demands at least two 'periods' for the DMA to work correctly
+	 * Note: period_bytes == 0 can only happen for firmware or library
+	 * loading, the binary sizes have the necessary alignment requirements
+	 * met
+	 */
+	if (!period_bytes) {
+		unsigned int chunk_size;
+
+		chunk_size = snd_sgbuf_get_chunk_size(dmab, 0, hstream->bufsize);
+
 		period_bytes = hstream->bufsize;
+		/*
+		 * If the memory is a single continuous area, force it to be
+		 * split into two 'periods', otherwise the transfer will be
+		 * split to multiple BDLE for each chunk in hda_setup_bdle()
+		 */
+		if (chunk_size == hstream->bufsize)
+			period_bytes /= 2;
+	}
```

Comment:
The other open I have is that the number of periods comes from topology, and I don't think we've ever had a solution with a single period.

So the question is: what does this commit actually fix?

Reply:
With audio we _always_ have at least two periods. The only case when we don't is the firmware and library loading.

Reply:
Hrm, we could have periods_min set to 1... Interesting, that will not work for sure.
This patch only handles the firmware/library loading (when hstream->period_bytes is set to 0).
PCM with periods=1 cannot work...